## Практическое домашнее задание на проверку знаний по модулю 1.


##### Тема тестирования - базовый эксплоративный анализ и аналитическое мышление. Вам будет предоставлен датасет и 25 вопросов касаемо данного датасета. Ваша задача - подумать как найти ответы на поставленные вопросы ниже. За каждый правильно оформленный и предоставленный ответ начисляется *4 балла*.

#### Ответ на решение присылать ссылкой на ваш гитхаб репозиторий на почту maratmovlamov2017@gmail.com

#### Сроки выполнения практического домашнего задания - ровно 7 календарных дней (168 часов) с момента публикации в группе. Прошу приступать к решению задач порционно. Прошу повторить дополнительные материалы, предоставленные в качестве самостоятельного изучения в конце презентаций, они вам понадобятся.



#### Критерии оценки присланного решения:

* Правильность полученного ответа - на каждый поставленный вопрос есть один и только один правильный ответ. В случае неправильного ответа на поставленный вопрос, предоставленный ответ на него считается проваленным, даже если код написан правильно и красиво. Общее количество баллов, которые можно получить ответив на все вопросы верно - 25 * 4 = 100 баллов, которые шкалируются к 10-ти бальной шкале.

___

* Аккуратность, правильность и читаемость кода - старайтесь писать код без повторного использования методов и ненужных переменных (помните про принцип DRY - don't repeat yourself). Плюсом будет, если ваш код будет упакован в лаконичную и красиво написанную функцию. Если вся работа выполненна аккуратно, без спагетти кода, к общему результату работы начисляется 1 балл.

---

* На предоставленные вопросы, дописывается комментарий для проверки вашего мышления и поиска ответа в виде Markdown комментария. В случае если комментарий на поставленный вопрос не предоставлен снимается 0.1 балл за каждый такой вопрос. То есть общий штраф за отсутсвие комментариев по всем вопросам 25 * -0.1 =  -2.5 балла.


#### Удачи и да прибудет с Вами сила :)



In [67]:
import pandas as pd
import numpy as np

In [68]:
df = pd.read_csv("./dataset/Spotify_Youtube.csv")

In [69]:
df.head()

Unnamed: 0    Artist                                        Url_spotify   
0           0  Gorillaz  https://open.spotify.com/artist/3AA28KZvwAUcZu...  \
1           1  Gorillaz  https://open.spotify.com/artist/3AA28KZvwAUcZu...   
2           2  Gorillaz  https://open.spotify.com/artist/3AA28KZvwAUcZu...   
3           3  Gorillaz  https://open.spotify.com/artist/3AA28KZvwAUcZu...   
4           4  Gorillaz  https://open.spotify.com/artist/3AA28KZvwAUcZu...   

                                           Track   
0                                 Feel Good Inc.  \
1                                Rhinestone Eyes   
2  New Gold (feat. Tame Impala and Bootie Brown)   
3                             On Melancholy Hill   
4                                 Clint Eastwood   

                                           Album Album_type   
0                                     Demon Days      album  \
1                                  Plastic Beach      album   
2  New Gold (feat. Tame Impala and Bootie Brown)     single   
3                                  Plastic Beach      album   
4                                       Gorillaz      album   

                                    Uri  Danceability  Energy   Key  ...   
0  spotify:track:0d28khcov6AiegSCpG5TuT         0.818   0.705   6.0  ...  \
1  spotify:track:1foMv2HQwfQ2vntFf9HFeG         0.676   0.703   8.0  ...   
2  spotify:track:64dLd6rVqDLtkXFYrEUHIU         0.695   0.923   1.0  ...   
3  spotify:track:0q6LuUqGLUiCPP1cbdwFs3         0.689   0.739   2.0  ...   
4  spotify:track:7yMiX7n9SBvadzox8T5jzT         0.663   0.694  10.0  ...   

                                   Url_youtube   
0  https://www.youtube.com/watch?v=HyHNuVaZJ-k  \
1  https://www.youtube.com/watch?v=yYDmaexVHic   
2  https://www.youtube.com/watch?v=qJa-VFwPpYA   
3  https://www.youtube.com/watch?v=04mfKJWDSzI   
4  https://www.youtube.com/watch?v=1V_xRb0x9aw   

                                               Title   Channel        Views   
0         Gorillaz - Feel Good Inc. (Official Video)  Gorillaz  693555221.0  \
1  Gorillaz - Rhinestone Eyes [Storyboard Film] (...  Gorillaz   72011645.0   
2  Gorillaz - New Gold ft. Tame Impala & Bootie B...  Gorillaz    8435055.0   
3     Gorillaz - On Melancholy Hill (Official Video)  Gorillaz  211754952.0   
4         Gorillaz - Clint Eastwood (Official Video)  Gorillaz  618480958.0   

       Likes  Comments                                        Description   
0  6220896.0  169907.0  Official HD Video for Gorillaz' fantastic trac...  \
1  1079128.0   31003.0  The official video for Gorillaz - Rhinestone E...   
2   282142.0    7399.0  Gorillaz - New Gold ft. Tame Impala & Bootie B...   
3  1788577.0   55229.0  Follow Gorillaz online:\nhttp://gorillaz.com \...   
4  6197318.0  155930.0  The official music video for Gorillaz - Clint ...   

   Licensed official_video        Stream  
0      True           True  1.040235e+09  
1      True           True  3.100837e+08  
2      True           True  6.306347e+07  
3      True           True  4.346636e+08  
4      True           True  6.172597e+08  

[5 rows x 28 columns]

### Структура датасета:

----

26 признаков, описание которых следует ниже:



- **Track**: name of the song, as visible on the Spotify platform.
- **Artist**: name of the artist.
- **Url_spotify**: the Url of the artist.
- **Album**: the album in wich the song is contained on Spotify.
- **Album_type**: indicates if the song is relesead on Spotify as a single or contained in an album.
- **Uri**: a spotify link used to find the song through the API.
- **Danceability**: describes how suitable a track is for dancing based on a combination of musical elements including tempo, rhythm stability, beat strength, and overall regularity. A value of 0.0 is least danceable and 1.0 is most danceable.
- **Energy:** is a measure from 0.0 to 1.0 and represents a perceptual measure of intensity and activity. Typically, energetic tracks feel fast, loud, and noisy. For example, death metal has high energy, while a Bach prelude scores low on the scale. Perceptual features contributing to this attribute include dynamic range, perceived loudness, timbre, onset rate, and general entropy.
- **Key**: the key the track is in. Integers map to pitches using standard Pitch Class notation. E.g. 0 = C, 1 = C♯/D♭, 2 = D, and so on. If no key was detected, the value is -1.
- **Loudness**: the overall loudness of a track in decibels (dB). Loudness values are averaged across the entire track and are useful for comparing relative loudness of tracks. Loudness is the quality of a sound that is the primary psychological correlate of physical strength (amplitude). Values typically range between -60 and 0 db.
- **Speechiness**: detects the presence of spoken words in a track. The more exclusively speech-like the recording (e.g. talk show, audio book, poetry), the closer to 1.0 the attribute value. Values above 0.66 describe tracks that are probably made entirely of spoken words. Values between 0.33 and 0.66 describe tracks that may contain both music and speech, either in sections or layered, including such cases as rap music. Values below 0.33 most likely represent music and other non-speech-like tracks.
- **Acousticness**: a confidence measure from 0.0 to 1.0 of whether the track is acoustic. 1.0 represents high confidence the track is acoustic.
- **Instrumentalness**: predicts whether a track contains no vocals. "Ooh" and "aah" sounds are treated as instrumental in this context. Rap or spoken word tracks are clearly "vocal". The closer the instrumentalness value is to 1.0, the greater likelihood the track contains no vocal content. Values above 0.5 are intended to represent instrumental tracks, but confidence is higher as the value approaches 1.0.
- **Liveness**: detects the presence of an audience in the recording. Higher liveness values represent an increased probability that the track was performed live. A value above 0.8 provides strong likelihood that the track is live.
- **Valence**: a measure from 0.0 to 1.0 describing the musical positiveness conveyed by a track. Tracks with high valence sound more positive (e.g. happy, cheerful, euphoric), while tracks with low valence sound more negative (e.g. sad, depressed, angry).
- **Tempo**: the overall estimated tempo of a track in beats per minute (BPM). In musical terminology, tempo is the speed or pace of a given piece and derives directly from the average beat duration.
- **Duration_ms**: the duration of the track in milliseconds.
- **Stream**: number of streams of the song on Spotify.
- **Url_youtube**: url of the video linked to the song on Youtube, if it have any.
- **Title**: title of the videoclip on youtube.
- **Channel**: name of the channel that have published the video.
- **Views**: number of views.
- **Likes**: number of likes.
- **Comments**: number of comments.
- **Description**: description of the video on Youtube.
- **Licensed**: Indicates whether the video represents licensed content, which means that the content was uploaded to a channel linked to a YouTube content partner and then claimed by that partner. official_video: boolean value that indicates if the video found is the official video of the song.

### Вопросы:

#### 1. Какая песня является наиболее транслируемой в датасете (признак Stream)?

In [70]:
df[df['Stream'] == df['Stream'].max()][['Artist', 'Track', 'Stream']]

Artist            Track        Stream
15250  The Weeknd  Blinding Lights  3.386520e+09

Наиболее транслируемой песней в датасете является песня Blinding Lights by The Weeknd

#### 2. У какого исполнителя больше всего просмотров на YouTube?

In [71]:
pd.pivot_table(data = df, index = 'Artist', values = 'Views', aggfunc=np.sum).sort_values(by='Views', ascending=False).head()

Views
Artist                    
Ed Sheeran    1.546021e+10
CoComelon     1.460167e+10
Katy Perry    1.312063e+10
Charlie Puth  1.216759e+10
Luis Fonsi    1.162811e+10

Исполнителем с наибольшим количеством просмотров на YouTube является Ed Sheeran

#### 3. Какова средняя продолжительность всех песен в минутах?

In [72]:
np.round(df['Duration_ms'].mean() / 60000, 4)

3.7453

Средняя продолжительность всех песен в минутах --- 3.7453 минут

#### 4. Какая нота (Key) чаще всего используется в песнях?

In [73]:
df['Key'].value_counts()

Key
0.0     2305
7.0     2252
1.0     2211
2.0     2021
9.0     1979
5.0     1731
11.0    1667
4.0     1515
8.0     1483
6.0     1443
10.0    1439
3.0      670
Name: count, dtype: int64

Чаще всего используется нота C = 0 (2305 раз)

#### 5. Какая самая популярная песня на YouTube?

In [74]:
df[df['Views'] == df['Views'].max()][['Artist', 'Track', 'Views']]

Artist      Track         Views
1147  Luis Fonsi  Despacito  8.079649e+09

Cамой популярной песней на YouTube является песня Despacito by Luis Fonsi

#### 6. У какой песни самый высокий балл за танцевальность (danceability )?

In [75]:
df[df['Danceability'] == df['Danceability'].max()][['Artist', 'Track', 'Danceability']]

Artist             Track  Danceability
750    Timbaland     Give It To Me         0.975
16555        BIA  CAN'T TOUCH THIS         0.975
18822       Dave      Funky Friday         0.975

Самый высокий балл за танцевальность наблюдается у песен CAN'T TOUCH THIS by BIA, Give It To Me by Timbaland, Funky Friday by Dave с показателем 0.975

#### 7. У какой песни самый высокий показатель энергии (energy score)?

In [76]:
df[df['Energy'] == df['Energy'].max()][['Artist', 'Track', 'Energy']]

Artist                          Track  Energy
20581  Rain Fruits Sounds   Rain and Thunderstorm, Pt. 3     1.0
20582  Rain Fruits Sounds   Rain and Thunderstorm, Pt. 4     1.0
20585  Rain Fruits Sounds   Rain and Thunderstorm, Pt. 7     1.0
20586  Rain Fruits Sounds   Rain and Thunderstorm, Pt. 6     1.0
20587  Rain Fruits Sounds  Rain and Thunderstorm, Pt. 33     1.0
20652      Sleepy Penguin          Gentle Piano Melodies     1.0

Самый высокий показатель энергии наблюдается у песен Rain and Thunderstorm, Pt. 7, 33, 4, 3, 6 by Rain Fruits Sounds и у песни Gentle Piano Melodies by Sleepy Penguin

#### 8. Какая песня имеет наивысший балл valence (valence score)?

In [77]:
df[df['Valence'] == df['Valence'].max()][['Artist', 'Track', 'Valence']]

Artist         Track  Valence
18514  Toby Fox  Spider Dance    0.993

Наивысшим баллом "valence" обладает песня Spider Dance by Toby Fox

#### 9. Будут ли песни с более высокой «valence» (музыкальной позитивностью) более популярны как на Spotify, так и на YouTube? Чтобы ответить на этот вопрос, подсчитайте среднюю оценку «valence» для песен в верхнем квартиле «Streams» и «Views» и сравните ее с общей средней «valence».

In [78]:
df_views = df.nlargest(int(len(df) * 0.25), columns = 'Views')
df_stream = df.nlargest(int(len(df) * 0.25), columns = 'Stream')
print(f"Средняя 'valence' в верхнем квартиле Spotify: {df_stream['Valence'].mean()}\n"
      f"Средняя 'valence' в верхнем квартиле YouTube: {df_views['Valence'].mean()}\n"
      f"Общая средняя 'valence': {df['Valence'].mean()}")

Средняя 'valence' в верхнем квартиле Spotify: 0.5235957377365779
Средняя 'valence' в верхнем квартиле YouTube: 0.5470677157752463
Общая средняя 'valence': 0.5298533240007725


Песни с более высокой valence более популярны на YouTube

#### 10. Какой канал на YouTube имеет наибольшее количество просмотров?

In [79]:
pd.pivot_table(data = df, index = 'Channel', values = 'Views', aggfunc=np.sum).sort_values(by='Views', ascending=False)

Views
Channel                                  
T-Series                     4.796159e+10
LuisFonsiVEVO                2.170252e+10
CalvinHarrisVEVO             1.828114e+10
Macklemore                   1.700341e+10
Ed Sheeran                   1.691822e+10
...                                   ...
Giada Coronas                3.100000e+01
Christian Rode - Topic       2.800000e+01
water Chillon                2.600000e+01
Bernhard Kellermann - Topic  8.000000e+00
Adi Mira Michaels - Topic    2.000000e+00

[6714 rows x 1 columns]

T-Series --- канал с наибольшим количеством просмотров

#### 11. У песни какого исполнителя больше всего комментариев на YouTube?

In [80]:
df[df['Comments'] == df['Comments'].max()][['Artist', 'Track', 'Comments']]

Artist     Track    Comments
14561    BTS  Dynamite  16083138.0

Наибольшим количеством комментариев обладает группа BTS

#### 12. Какой самый распространенный темп (tempo) песен?

In [81]:
df['Tempo'].value_counts()

Tempo
77.986     24
106.002    19
0.000      17
120.031    12
119.982    11
           ..
100.217     1
165.313     1
82.525      1
82.499      1
155.378     1
Name: count, Length: 15024, dtype: int64

Самый распространенный темп --- 77.986 (24 наблюдения)

#### 13. Какая песня имеет самый высокий показатель по speechiness?

In [82]:
df[df['Speechiness'] == df['Speechiness'].max()][['Artist', 'Track', 'Speechiness']]

Artist   
5158  Sir Arthur Conan Doyle  \

                                                  Track  Speechiness  
5158  Teil 7 - Sherlock Holmes und der blinde Bettle...        0.964

Самый высокий показатель speechiness имеет песня Teil 7 - Sherlock Holmes und der blinde Bettler - Die neuen Abenteuer, Folge 50 (по своей сути часть аудиокниги)

#### 14. Какая самая громкая песня в наборе данных?

In [83]:
df[df['Loudness'] == df['Loudness'].max()][['Artist', 'Track', 'Loudness']]

Artist  Track  Loudness
9423  Sub Focus  Siren      0.92

Самая громкая песня в наборе данных --- Siren by Sub Focus

#### 15. У какого артиста самый высокий средний балл danceability?

In [84]:
pd.pivot_table(data = df, index= 'Artist', values= 'Danceability', aggfunc= np.mean).sort_values(by = 'Danceability', ascending=False).head(1)

Danceability
Artist                
Saweetie        0.8858

Самый высокий балл danceability имеет артист Saweetie с показателем 0.8858

#### 16. Как средний показатель «speechiness» песен зависит от «Album_type»? Песни, выпущенные как синглы, чаще содержат больше слов согласно анализу?

In [85]:
pd.pivot_table(data = df, index = 'Album_type', values = 'Speechiness', aggfunc=np.mean)

Speechiness
Album_type              
album           0.094900
compilation     0.078867
single          0.103867

Песни, выпущенные как синглы, содержат больше всего слов, далее выпущенные в рамках альбома, треки выпущенные как компиляция имеют меньше всего слов

#### 17. Каково среднее количество прослушиваний песен, которые входят в топ 10% лучших по метрике танцевальности?

In [86]:
df_dance = df.nlargest(int(len(df) * 0.1), columns = 'Danceability')
df_dance['Stream'].mean()

151937856.8966193

Среднее количество прослушиваний для песен, которые входят в топ 10% лучших по метрике танцевальности, --- 151937856.8966193

#### 18. Для песен, которые входят в топ 10% лучших с точки зрения «danceability» метрики, как соотносится их средний показатель «tempo» со средним показателем «tempo» всех песен?

In [87]:
print(f"Средний показатель tempo для 10% лучших с точки зрения 'danceability': {df_dance['Tempo'].mean()}\n"
      f"Средний общий показатель tempo: {df['Tempo'].mean()}" )

Средний показатель tempo для 10% лучших с точки зрения 'danceability': 115.05771607918878
Средний общий показатель tempo: 120.63833993048851


Песни, входящие в топ 10% лучших с точки зрения танцевальности, обладают более низким темпом, чем средний темп для всех песен

#### 19. Какова корреляция между «valence» метрикой песни и ее популярностью на YouTube (по показателю «Views») и Spotify (по показателю «Stream»)? Счастливые песни более популярны? Valency отвечает за жизнерадостность песни.

In [88]:
df[['Valence', 'Views', 'Stream']].corr(numeric_only=True)

Valence     Views    Stream
Valence  1.000000  0.035423 -0.012109
Views    0.035423  1.000000  0.601905
Stream  -0.012109  0.601905  1.000000

Показатель valence имеет очень низкую корреляцию с показателями Views и Stream, поэтому можно сделать вывод, что счастливые песни не являются более популярными

#### 20. Какова средняя громкость песен, которые входят в топ 10% лучших по энергетике (Energy)?

In [89]:
df_energy = df.nlargest(int(len(df) * 0.1), columns = 'Energy')
df_energy['Loudness'].mean()

-4.630413809753742

Средняя громкость песен, входящих в топ 10% по энергетике --- -4.630413809753742

#### 21. У какого исполнителя больше всего песен, которые входят в топ-10% как по просмотрам на YouTube, так и по стримам Spotify?

In [90]:
df_views = df.nlargest(int(len(df) * 0.1), columns = 'Views')
df_stream = df.nlargest(int(len(df) * 0.1), columns = 'Stream')
df_inner = pd.merge(df_views, df_stream, on = 'Unnamed: 0', how = 'inner',  suffixes=('', '_DROP')).filter(regex='^(?!.*_DROP)')
df_inner['Artist'].value_counts()


Artist
Calvin Harris    9
Dua Lipa         9
Coldplay         9
Ariana Grande    9
J Balvin         8
                ..
Kodak Black      1
AURORA           1
The Killers      1
Vance Joy        1
Soundgarden      1
Name: count, Length: 480, dtype: int64

Исполнителями, которые имеют наибольшее количество песен одновременно входящих в топ 10% по просмотрам на YouTube и по стримам на Spotify, являются: Calvin Harris, Dua Lipa, Coldplay и Ariana Grande --- по 9 песен

#### 22. Какова корреляция между «energy» песни и ее «loudness»?

In [91]:
round(df['Energy'].corr(df['Loudness']), 4)

0.7448

Корреляция между energy и loudness равна 0,7448 --- сильная корреляция

#### 23. Каково среднее количество лайков на YouTube для песен, которые находятся в верхнем квартиле (25%) и как это соотносится с общим средним показателем?

In [92]:
df_likes = df.nlargest(int(len(df) * 0.25), columns = 'Likes')
print(f"Среднее количество лайков для песен в верхнем квартиле: {df_likes['Likes'].mean()}\nСреднее количество лайков для всех песен: {df['Likes'].mean()}")

Среднее количество лайков для песен в верхнем квартиле: 2260833.73991118
Среднее количество лайков для всех песен: 663341.0613074292


Среднее количество лайков для песен в верхнем квартиле ~ больше на 1.5 млн, чем общий показатель

#### 24. Есть ли существенная разница в средней «valence» метрикой для песен, выпущенных как синглы, по сравнению с песнями, выпущенными в виде альбома?

In [93]:
pd.pivot_table(data=df, index= 'Album_type', values = 'Valence', aggfunc=np.mean )

Valence
Album_type           
album        0.529771
compilation  0.529679
single       0.530126

Существенной разницы по средней valence в зависимости от типа альбома нет

#### 25. Каково среднее количество «комментариев» на YouTube для песен с показателем «speechiness» выше 0,66 (что указывает на большое количество произнесенных слов) и как это соотносится с общим средним показателем?

In [94]:
print(f"Среднее количество комментариев для песен с показателем 'speechies'>0.66: {df[df['Speechiness']>0.66]['Comments'].mean()}\nСреднее количество комментариев в целом: {df['Comments'].mean()}")

Среднее количество комментариев для песен с показателем 'speechies'>0.66: 334.60550458715596
Среднее количество комментариев в целом: 27518.993696957667


Среднее количество комментариев для песен с показателем speechiness > 0.66 (преимущественно аудиокниги) гораздо ниже чем общий показатель